src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2172,10 +2172,944 @@ - (void)doTestSubscriptionPoolWithSize:(NSInteger)subscriptionPoolSize
+                        XCTAssertTrue([device _persistedClusterContains:path]);
+                        XCTAssertNotNil([device _getPersistedClusterDataForPath:path]);
+                    } else {
+                        XCTAssertNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                        XCTAssertFalse([device _persistedClusterContains:path]);
+                        XCTAssertNil([device _getPersistedClusterDataForPath:path]);
+                    }
+                }
+            }
+        });
+        [reportEndExpectation fulfill];
+    };
+
+    [device unitTestInjectAttributeReport:attributeReport fromSubscription:YES];
+
+    [self waitForExpectations:@[ attributeDataReceivedExpectation, reportEndExpectation ] timeout:60];
+
+    [controller.controllerDataStore clearAllStoredClusterData];
+    NSDictionary * storedClusterDataAfterClear = [controller.controllerDataStore getStoredClusterDataForNodeID:deviceID];
+    XCTAssertEqual(storedClusterDataAfterClear.count, 0);
+
+    [controller removeDevice:device];
+    // Reset our commissionee.
+    __auto_type * baseDevice = [MTRBaseDevice deviceWithNodeID:deviceID controller:controller];
+    ResetCommissionee(baseDevice, queue, self, kTimeoutInSeconds);
+
+    [controller shutdown];
+    XCTAssertFalse([controller isRunning]);
+}
+
+- (void)test015_TestDataStoreMTRDeviceDoNotPruneOrphanedClusters
+{
+    __auto_type * factory = [MTRDeviceControllerFactory sharedInstance];
+    XCTAssertNotNil(factory);
+
+    __auto_type queue = dispatch_get_main_queue();
+
+    __auto_type * rootKeys = [[MTRTestKeys alloc] init];
+    XCTAssertNotNil(rootKeys);
+
+    __auto_type * operationalKeys = [[MTRTestKeys alloc] init];
+    XCTAssertNotNil(operationalKeys);
+
+    NSNumber * nodeID = @(123);
+    NSNumber * fabricID = @(456);
+
+    NSError * error;
+    __auto_type * storageDelegate = [[MTRTestPerControllerStorageWithBulkReadWrite alloc] initWithControllerID:[NSUUID UUID]];
+    MTRPerControllerStorageTestsCertificateIssuer * certificateIssuer;
+    MTRDeviceController * controller = [self startControllerWithRootKeys:rootKeys
+                                                         operationalKeys:operationalKeys
+                                                                fabricID:fabricID
+                                                                  nodeID:nodeID
+                                                                 storage:storageDelegate
+                                                                   error:&error
+                                                       certificateIssuer:&certificateIssuer];
+    XCTAssertNil(error);
+    XCTAssertNotNil(controller);
+    XCTAssertTrue([controller isRunning]);
+
+    XCTAssertEqualObjects(controller.controllerNodeID, nodeID);
+
+    // Now commission the device, to test that that works.
```

Comment:
See comments above; please remove all the irrelevant bits from these tests.

Also, please have a helper method for "yeah, whatever, just set things up so I can get an MTRDevice, please", so we don't have so many copies of that code.

Reply:
done.